### PR TITLE
[mlir][spirv] Add folding for SNegate, [Logical]Not

### PR DESCRIPTION
**Pull request by inbelic**
Add missing constant propogation folder for SNegate, [Logical]Not.

Implement additional folding when !(!x) for all ops.

This helps for readability of lowered code into SPIR-V.

Part of work for #70704

### CODE REVIEW

File: mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp [hunk 1]
```diff
@@ -643,6 +643,45 @@ OpFoldResult spirv::UModOp::fold(FoldAdaptor adaptor) {
...
 
+//===----------------------------------------------------------------------===//
+// spirv.SNegate
+//===----------------------------------------------------------------------===//
+
+OpFoldResult spirv::SNegateOp::fold(FoldAdaptor adaptor) {
+  // -(-x) = 0 - (0 - x) = x
+  auto op = getOperand();
+  if (auto negateOp = op.getDefiningOp<spirv::SNegateOp>())
+    return negateOp->getOperand(0);
```
Comment (kuhar) [inline mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:654]: I wasn't sure if this works for INT_MIN, it should check out like that:
- `- INT_MIN == INT_MIN`
- `- (- INT_MIN) == - INT_MIN = INT_MIN`

which looks fine

File: mlir/test/Dialect/SPIRV/Transforms/canonicalize.mlir [hunk 2]
```diff
@@ -1006,6 +1006,90 @@ func.func @umod_fail_fold(%arg0: i32) -> (i32, i32) {
...
+
+// CHECK-LABEL: @const_fold_vector_snegate
+func.func @const_fold_vector_snegate() -> vector<3xi32> {
+  // CHECK: spirv.Constant dense<[0, 3, -3]>
+  %cv = spirv.Constant dense<[0, -3, 3]> : vector<3xi32>
+  %0 = spirv.SNegate %cv : vector<3xi32>
+  return %0  : vector<3xi32>
+}
+
+// -----
```
Comment (kuhar) [inline mlir/test/Dialect/SPIRV/Transforms/canonicalize.mlir:1048]: Can we have one test with INT_MIN? We can do it over i8 to keep the constants small.

I would expect `spirv.SNegate -128 ==> spirv.Constant -128 : i8`

**Review by kuhar — COMMENTED** [review]
Looks great, I think it would be worth adding one more test for though.

File: mlir/test/Dialect/SPIRV/Transforms/canonicalize.mlir (hunk 2 above)
Comment (inbelic) [reply mlir/test/Dialect/SPIRV/Transforms/canonicalize.mlir:1060]: Good catch, it is not immediately clear so I agree the testcase is important

**Review by kuhar — APPROVED** [review]
LGTM

**Pull request merged**